src/runner.py

```diff
@@ -88,596 +47,416 @@ class LxdInstanceConfig(TypedDict):
 
-class RunnerRemoveError(RunnerError):
-    """Error for runner removal failure."""
+            try:
+                self.instance.delete(wait=True)
+            except Exception:  # nosec B110
+                # Ephemeral containers should auto-delete when stopped;
+                # this is just a fall-back.
+                pass
+        else:
+            # We somehow have a non-running instance which should have been
+            # ephemeral. Try to delete it and allow any errors doing so to
+            # surface.
+            try:
+                self.instance.delete(wait=True)
+            except Exception as err:
+                raise RunnerRemoveError() from err
 
+    @retry(tries=10, delay=1, logger=logger)
+    def _create_instance(
+        self, image: str, resources: VirtualMachineResources, ephemeral: bool = True
+    ) -> pylxd.models.Instance:
+        """Create a instance of runner.
 
-class RunnerStartError(RunnerError):
-    """Error for runner start failure."""
+        Args:
+            image: Image to launch the instance hosting the runner.
+            resources: Configuration of the virtual machine resources.
+            ephemeral: Whether the instance is ephemeral. Defaults to True.
 
+        Returns:
+            LXD instance of the runner.
+        """
+        logger.info("Creating an LXD instance for runner: %s", self.name)
 
-class RunnerManager:
-    """Class for querying and controlling the runners.
+        self._ensure_runner_profile()
+        resource_profile = self._get_resource_profile(resources)
 
-    Attributes:
-        proxies (Dict[str, str]): Mapping of proxy env vars.
-        path (str): GitHub repository path in the format '<org>/<repo>', or the GitHub org name.
-        app_name (str): An name for the set of runners.
-        reconcile_interval (int): Number of minutes between each reconciliation of runners state.
-    """
+        # Create runner instance.
+        instance_config: LxdInstanceConfig = {
+            "name": self.name,
+            "type": "virtual-machine",
+            "source": {
+                "type": "image",
+                "mode": "pull",
+                "server": "https://cloud-images.ubuntu.com/daily",
+                "protocol": "simplestreams",
+                "alias": image,
+            },
+            "ephemeral": ephemeral,
+            "profiles": ["default", "runner", resource_profile],
+        }
 
-    # TODO: Verify if this violates bandit B108.
-    runner_bin_path = Path("/var/cache/github-runner-operator/runner.tgz")
-    env_file = Path("/opt/github-runner/.env")
+        instance = self._lxd.instances.create(config=instance_config, wait=True)
+        self.exist = True
+        return instance
 
-    def __init__(self, path: str, token: str, app_name: str, reconcile_interval: int) -> None:
-        """Construct RunnerManager object for creating and managing runners.
+    @retry(tries=10, delay=1, logger=logger)
```

Comment:
We also have here embedded retries, possible to refactor to avoid this situation ?